models/eprop_readout.cpp

```diff
@@ -0,0 +1,388 @@
+
+eprop_readout::eprop_readout()
+  : EpropArchivingNodeReadout()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+eprop_readout::eprop_readout( const eprop_readout& n )
+  : EpropArchivingNodeReadout( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+eprop_readout::init_buffers_()
+{
+  B_.spikes_.clear();   // includes resize
+  B_.currents_.clear(); // includes resize
+  B_.logger_.reset();   // includes resize
+}
+
+void
+eprop_readout::pre_run_hook()
+{
+  B_.logger_.init(); // ensures initialization in case multimeter connected after Simulate
+
+  compute_error_signal = &eprop_readout::compute_error_signal_mean_squared_error;
+
+  const double dt = Time::get_resolution().get_ms();
+
+  V_.P_v_m_ = std::exp( -dt / P_.tau_m_ );
+  V_.P_i_in_ = P_.tau_m_ / P_.C_m_ * ( 1.0 - V_.P_v_m_ );
+  V_.P_z_in_ = P_.regular_spike_arrival_ ? 1.0 : 1.0 - V_.P_v_m_;
+}
+
+long
+eprop_readout::get_shift() const
+{
+  return offset_gen_ + delay_in_rec_;
+}
+
+bool
+eprop_readout::is_eprop_recurrent_node() const
+{
+  return false;
+}
+
+/* ----------------------------------------------------------------
+ * Update function
+ * ---------------------------------------------------------------- */
+
+void
+eprop_readout::update( Time const& origin, const long from, const long to )
+{
+  const size_t buffer_size = kernel().connection_manager.get_min_delay();
+
+  std::vector< double > error_signal_buffer( buffer_size, 0.0 );
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+    const long t = origin.get_steps() + lag;
+
+    S_.z_in_ = B_.spikes_.get_value( lag );
+
+    S_.v_m_ = V_.P_i_in_ * S_.i_in_ + V_.P_z_in_ * S_.z_in_ + V_.P_v_m_ * S_.v_m_;
+    S_.v_m_ = std::max( S_.v_m_, P_.V_min_ );
+
+    ( this->*compute_error_signal )( lag );
+
+    S_.target_signal_ *= S_.learning_window_signal_;
+    S_.readout_signal_ *= S_.learning_window_signal_;
+    S_.error_signal_ *= S_.learning_window_signal_;
+
+    error_signal_buffer[ lag ] = S_.error_signal_;
+
+    emplace_new_eprop_history_entry( t, false );
+
+    write_error_signal_to_history( t, S_.error_signal_, false );
+
+    S_.i_in_ = B_.currents_.get_value( lag ) + P_.I_e_;
+
+    B_.logger_.record_data( t );
+  }
+
+  LearningSignalConnectionEvent error_signal_event;
+  error_signal_event.set_coeffarray( error_signal_buffer );
+  kernel().event_delivery_manager.send_secondary( *this, error_signal_event );
+
+  return;
+}
+
+/* ----------------------------------------------------------------
+ * Error signal functions
+ * ---------------------------------------------------------------- */
+
+void
+eprop_readout::compute_error_signal_mean_squared_error( const long lag )
+{
+  S_.readout_signal_ = S_.v_m_ + P_.E_L_;
+  S_.error_signal_ = S_.readout_signal_ - S_.target_signal_;
```

Comment:
Why is this method called `..._mean_squared_error()`?  I do not see any square here. Applies also to the `bsshslm_2020` variant.

Reply:
The detail here is that the error signal carries information about the derivative of the loss function rather than the loss function itself

Reply:
OK.